### PR TITLE
Rework request cancellation

### PR DESCRIPTION
**Pull request by arthurschreiber**
This is an attempt to rework how request cancellation works, to make it more generic (it now works with regular requests and both streaming as well as non-streaming bulk loads) and more conformant to the MS-TDS spec.

This also adds a new `.cancel` method on the `Request` and `BulkLoad` classes, and deprecates the `cancel` method on the `Connection` class itself.

The changes in 8c9ce1d warrant a major version bump, as they change the behaviour of requests that are canceled after they were sent to the server. Instead of continuing to forward various events from the token stream parser to the canceled request, all this information will be "discarded", as described in [`3.2.5.8 Sent Attention State`](https://msdn.microsoft.com/en-us/library/dd340677.aspx) of the MS-TDS specification:

> If the response is structurally valid and it does not acknowledge the Attention as described in section 2.2.1.7, then the TDS client MUST discard any data contained in the response and remain in the "Sent Attention" state.
> 
> If the response is structurally valid and it acknowledges the Attention as described in section 2.2.1.7, then the TDS client MUST discard any data contained in the response, indicate the completion of the query to the upper layer together with the cause of the Attention (either an upper-layer cancellation as described in section 3.2.4 or query timeout as described in section 3.2.2), and enter the "Logged In" state.

Fixes: https://github.com/tediousjs/tedious/issues/545
Fixes: https://github.com/tediousjs/tedious/issues/534
Fixes: https://github.com/tediousjs/tedious/pull/566

### CODE REVIEW

File: src/connection.js [hunk 1]
```diff
@@ -1638,8 +1667,23 @@ class Connection extends EventEmitter {
       this.request.rows = [];
       this.request.rst = [];
 
+      let message;
+
+      this.request.once('cancel', () => {
```
Comment (chdh) [inline src/connection.js]: Somehow there should be a check whether the cancel event stems from the currently active request. When the user calls calcel on an old request or BulkLoad, it should be ignored.

File: src/connection.js [hunk 2]
```diff
@@ -1672,13 +1718,8 @@ class Connection extends EventEmitter {
       const message = 'Requests can only be canceled in the ' + this.STATE.SENT_CLIENT_REQUEST.name + ' state, not the ' + this.state.name + ' state';
```
Comment (chdh) [inline src/connection.js]: When cancel() is called after a timeout, a debug message is still displayed, see #545.

File: src/connection.js [hunk 3]
```diff
@@ -1638,8 +1667,23 @@ class Connection extends EventEmitter {
...
+      let message;
+
+      this.request.once('cancel', () => {
+        // There's two ways to handle request cancelation:
+        if (message.writable) {
+          // - if the message is still writable, we'll set the ignore bit
+          message.ignore = true;
+        } else {
+          // - but if the message has been ended (and thus has been fully sent off),
+          //   we need to send an `ATTENTION` message to the server
```
Comment (chdh) [inline src/connection.js]: Can the connection already be in the SENT_ATTENTION state, because `requesttimeout()` has executed?
Or in any other state?

**Comment by chdh** [comment]
I have made some comments, but the fundamental problem is that the program logic in `connection.js` is much too complicated to really comprehend it. There is such a jungle of events, callbacks, flags, states, etc.

The TDS protocol state machine should be encapsulated in a separate module and isolated from the rest.

File: src/connection.js (hunk 2 above)
Comment (arthurschreiber) [reply src/connection.js]: I want to deprecate `.cancel` on `Connection`, and thus did not touch this function more than necessary. We could change this function to:

```js
if (!this.request) {
  return false;
}

if (this.request.canceled) {
  return false;
}

this.request.cancel();
return true;
```

And still deprecate (and later remove) this function? 🤔

File: src/connection.js (hunk 3 above)
Comment (arthurschreiber) [reply src/connection.js]: You're right. 🙇 The request timeout functionality was still handling the sending of the attention message itself. I changed it to simply call `this.request.cancel` instead. This way, only the first call to `.cancel` will have any effect, and all later calls will not do anything.

File: src/connection.js (hunk 1 above)
Comment (arthurschreiber) [reply src/connection.js]: You're right! I'll rework this a bit to ensure we only ever handle `cancel` events from the current request. I'll also add some tests for this.

**Comment by arthurschreiber** [comment]
> I have made some comments, but the fundamental problem is that the program logic in `connection.js` is much too complicated to really comprehend it. There is such a jungle of events, callbacks, flags, states, etc.
> 
> The TDS protocol state machine should be encapsulated in a separate module and isolated from the rest. There should be clear documented APIs for every module. Private and public variables should be distinguished, etc.

Yeah, you're totally right. 😞 The way the state machine is currently translated to code is... not quite optimal. I'm trying to move the code in a better direction, through a small set of changes that I have in mind. The first set of changes is over https://github.com/tediousjs/tedious/pull/847. It's nothing revolutionary, but I feel like doing small, iterative changes is going to yield results as soon as possible and will help us move in the right direction quicker.

**Pull request merged**